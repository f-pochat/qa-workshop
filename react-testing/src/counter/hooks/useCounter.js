import { useCallback, useState } from 'react'

export const useCounter = (initialValue = 0) => {
  const [count, setCount] = useState(initialValue)

  const increment = useCallback(() => {
    setCount((c) => c + 1)
  }, [])

  const decrement = useCallback(() => {
    setCount((c) => c - 1)
  }, [])

  const reset = useCallback(() => {
    setCount(initialValue)
  }, [initialValue])

	return {
		count,
		setCount,
		increment,
		decrement,
		reset,
	}
}
