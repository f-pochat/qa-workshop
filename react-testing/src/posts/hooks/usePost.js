import axios from 'axios'
import { useQuery } from '@tanstack/react-query'

export const usePost = (id, options = {}) => {
	return useQuery({
    queryKey: ['post', id],
    queryFn: async () => {
      // ideally, this should be in a separate function
			const { data } = await axios.get(`https://jsonplaceholder.typicode.com/posts/${id}`)
      return data
    },
		...options,
		enabled: (options.enabled ?? true) && !!id
  })
}