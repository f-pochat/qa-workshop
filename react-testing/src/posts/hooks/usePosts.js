import axios from 'axios'
import { useQuery } from '@tanstack/react-query'

export const usePosts = (options = {}) => {
	return useQuery({
    queryKey: ['posts'],
    queryFn: async () => {
      // ideally, this should be in a separate function
      const { data } = await axios.get('https://jsonplaceholder.typicode.com/posts')
      return data
    },
    ...options
  })
}