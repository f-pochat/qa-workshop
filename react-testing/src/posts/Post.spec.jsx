// Task 2: test the Post component.
// Acceptance criteria
// 1) Include at least one snapshot.
// 2) The copy to clipboard functionality must be fully tested.

import React from 'react';
import { describe, it, expect } from 'vitest';
import { fireEvent, getByTestId, render, waitFor } from '@testing-library/react';
import { Post } from './Post';
import { MemoryRouter } from 'react-router-dom';
import { usePost } from './hooks';
import { successResponse } from './__mocks__';
import userEvent from '@testing-library/user-event';
import { useToast } from '@/components/ui/use-toast';

vi.mock('./hooks/usePost', () => ({
    usePost: vi.fn()
}))

vi.mock('../components/button', () => ({
    Button: <div></div>
}))


const mockClipboard = vi.fn(() => Promise.resolve());
Object.assign(navigator, {
    clipboard: {
        writeText: mockClipboard
    }
});

describe('Post', () => {
    it('renders correctly', () => {
        usePost.mockReturnValue({
            status: 'success',
            data: successResponse[0]
        })

        const { container } = render(<Post />, { wrapper: MemoryRouter });
        expect(container).toMatchSnapshot();
    });

    it('Test clipboard', async () => {
        const { container } = render(<Post />, { wrapper: MemoryRouter });

        const shareButton = container.querySelector("svg[id='copyIcon']")
        await userEvent.click(shareButton);
        expect(mockClipboard).toHaveBeenCalledWith(expect.stringContaining('http://localhost'));
    });
});
