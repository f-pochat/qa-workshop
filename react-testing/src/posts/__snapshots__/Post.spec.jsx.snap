// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Post > renders correctly 1`] = `
<div>
  <div
    class="flex items-center"
  >
    <a
      class="flex items-center"
      href="/posts"
    >
      <svg
        class="lucide lucide-arrow-big-left "
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M18 15h-6v4l-7-7 7-7v4h6v6z"
        />
      </svg>
      <span>
        Return to posts
      </span>
    </a>
    <button
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 w-10 ml-12 cursor-pointer"
    >
      <svg
        class="lucide lucide-share2 "
        fill="none"
        height="24"
        id="copyIcon"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="18"
          cy="5"
          r="3"
        />
        <circle
          cx="6"
          cy="12"
          r="3"
        />
        <circle
          cx="18"
          cy="19"
          r="3"
        />
        <line
          x1="8.59"
          x2="15.42"
          y1="13.51"
          y2="17.49"
        />
        <line
          x1="15.41"
          x2="8.59"
          y1="6.51"
          y2="10.49"
        />
      </svg>
    </button>
  </div>
  <div>
    <div>
      <h4
        class="text-lg font-semibold mb-2 text-center"
      >
        sunt aut facere repellat provident occaecati excepturi optio reprehenderit
      </h4>
      <p
        class="text-lg"
      >
        quia et suscipit
suscipit recusandae consequuntur expedita et cum
reprehenderit molestiae ut ut quas totam
nostrum rerum est autem sunt rem eveniet architecto
      </p>
    </div>
  </div>
</div>
`;
